### ISSUE TITLE
Is it theoretically possible to trace back how to modify values to produce a desired normal form?

### ISSUE THREAD

**Issue by changlinli** [dhall-lang/dhall-lang#626]
Not sure if this is a good place to ask these sorts of questions, but here goes...

An problem that happens sometimes when approaching pre-existing config files is figuring out the "proper" way to modify values.

In the case of Dhall, let's imagine that I have the normalized form
```
{ host = "magical-host"
, port = "9999"
-- and more stuff
}
```
created from a large collection of Dhall expressions. As a new developer I might want to change `magical-host` to `nonmagical-host`, but I might not know what the "proper" way of doing this is (e.g. maybe there's some `Constants.dhall` file I'm supposed to change or maybe `magical` and `host` are defined separately and then later concatenated together and I'm just supposed to change the `magical` part in some `HostsPrefix.dhall` file etc.).

If I can specify some constraints (e.g. you can only modify top-level definitions or you can only modify definitions in some file `foo`), and then a final desired normalized result (e.g. a change in a current field or adding a new element to a list), would it be possible in theory, given that everything always has a unique normal form, to trace back what changes (within the specified constraints) could give rise to a new normal form.

I ask this mainly as a theoretical concern to see whether such a system is possible to build on top of Dhall, as it would be extremely useful for understanding how to modify a pre-existing set of configuration files.

**Comment by singpolyma** [comment]
It is always possible to know what will happen to the normal form after a change by normalizing... but I'm not sure that's what you're asking.

**Comment by Gabriel439** [comment]
@singpolyma: I think what they mean is that they want to be able to trace the origin of all sub-expressions within the result so that they can know which file(s) to modify.

In other words, suppose they had the following Dhall files:

```dhall
-- ./foo.dhall
1
```

```dhall
-- ./result.dhall
{ foo = ./foo.dhall, bar = True }
```

... the latter of which produces this result:

```dhall
{ foo = 1, bar = True }
```

... and then they want to know which file they have to modify to change the output to:

```dhall
{ foo = 2, bar = True }
```

... then they are asking if the interpreter or IDE could tell them which files and source positions contributed to that subexpression.  In this case the answer would be that they have to edit either `./foo.dhall` or the first field of `./result.dhall`.

I think it should be possible in principle for the Haskell implementations since we do attach source spans to every sub-expression and the we could preserve all contributing source spans when β-normalizing expressions.

**Comment by changlinli** [comment]
@Gabriel439 Yep you've got it. The only added thing is that you probably want to be able to turn on some heuristics filters because for for large, complex configurations, there's probably a large number of files and source positions that contribute to a given subexpression and only a small handful of them are "reasonable" for a person to edit to get a new value.

Good to know that it seems possible for at least one implementation to do it. I'll take a closer look at the Haskell implementation then.

Thanks!

**Comment by sjakobi** [comment]
[This announcement](https://mail.haskell.org/pipermail/haskell-cafe/2020-February/131875.html) for a new `provenience` Haskell library reminded me of this issue.
